src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeFactory.cs

```diff
@@ -0,0 +1,421 @@
+    // Since the tree would be similar in both cases, we can later rewrite
+    // the whole thing as either a pattern or expression if needed.
+    internal static class AnalyzedNodeFactory
+    {
+        public static AnalyzedNode? Visit(IOperation operation)
+        {
+            return VisitCore(operation, input: null);
+        }
+
+        private static AnalyzedNode? VisitCore(IOperation operation, Evaluation? input)
+        {
+            Debug.Assert(input is null || operation is IPatternOperation || operation is IInvalidOperation);
+
+            return operation switch
+            {
+                IBinaryOperation op => VisitBinaryOperator(op),
+                IPatternOperation op => VisitPatternOperation(input, op),
+                IIsTypeOperation op => new Type(VisitInput(op.ValueOperand), op.TypeOperand, op.Syntax),
+                IIsPatternOperation op => VisitPatternOperation(VisitInput(op.Value), op.Pattern),
+                IPatternCaseClauseOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                ISwitchExpressionArmOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op => VisitNotOperator(op),
+                var op => VisitInput(op)
+            };
+        }
+
+        [return: NotNullIfNotNull("operation")]
+        private static Evaluation? VisitInput(IOperation? operation, Evaluation? input = null)
+        {
+            if (operation is null)
+                return null;
+
+            return VisitNode(operation, input) ?? new OperationEvaluation(operation);
+
+            static Evaluation? VisitNode(IOperation operation, Evaluation? input)
+            {
+                return operation switch
+                {
+                    IMemberReferenceOperation op when IsImplicitThisOrBaseMemberAccess(op) => null,
+                    ILocalReferenceOperation op => new Variable(input: null, op.Local, op.Syntax),
+                    IFieldReferenceOperation op => new MemberEvaluation(VisitInput(op.Instance, input), op),
+                    IPropertyReferenceOperation op when op.Property.IsIndexer => VisitIndexerReference(op),
+                    IPropertyReferenceOperation op => VisitNullableTypeMemberAccess(op) ??
+                                                      new MemberEvaluation(VisitInput(op.Instance, input), op),
+                    IConditionalAccessOperation op => VisitNode(op.WhenNotNull, VisitInput(op.Operation, input)),
+                    IConversionOperation op when op.Conversion.IsUserDefined => null,
+                    IConversionOperation op when op.Conversion.IsImplicit => VisitInput(op.Operand, input),
+                    IConversionOperation op => new Type(VisitInput(op.Operand, input), op.Type, op.Syntax),
+                    IConditionalAccessInstanceOperation => input,
+                    _ => null,
```

Comment:
honestly, all this can use docs.  even the simple stuff.  for example, i have no idea why `IConditionalAccessInstanceOperation => input` is appropriate.